contents/blog/how-we-improved-feature-flags-resiliency.md

```diff
@@ -0,0 +1,145 @@
+---
+date: 2023-03-28
+title: How we improved feature flag resiliency
+rootPage: /blog
+sidebar: Blog
+showTitle: true
+---
+
+Feature flags as a service is an interesting space to be working in. If your service stops working, not only does it affect your customers, but also your
+customer's customers, since they rely on you to make sure their app works fine.
+
+In contrast to say, the PostHog interface not loading, where the problem is constrained to your customers. It's not great, for sure, but it's better than event ingestion and feature flags going down.
+
+Further, flags can be very sensitive to latency: If it takes 5 seconds for your flags to evaluate, that is you holding up your customers application for 5 seconds. No, you can't wait to load them asynchronously as you need this result to determine what to show. Your business logic depends on this value.
+
+So, over the past quarter, one of our key goals has been to make feature flags fast and resilient: Even if the PostHog interface goes down, none of the SDKs that query flags should go down, nor the pure API interface. Further, for the strict latency flags, we need a way to resolve them in 50ms.
+
+This post is about how we did it, and what we learned along the way.
+
+Before we jump into how we improved things, it's worth listing out constraints that we can use to create solutions that otherwise wouldn't be possible.
+
+
+## Special Problem Constraints
+
+### PostHog Flags are deterministic
+
+Given a user identity, and a PostHog feature flag, the flag result is deterministic and sticky. So for example, if I roll out a flag to 30% of people, the person who's inside this window remains inside this window even if the rollout percentage goes up to 40, 80, or 100%.
+
+We guarantee this by computing a hash of the user id and the flag key.
+
+So, we never need a server to determine whether a flag belongs to a given rollout % or not. However, if a flag depends on person properties, we need to go to a database, check against known properties, and then determine if the flag matches or not.
+
+### Flags are evaluated multiple times in a session
+
+Flags are evaluated multiple times on a page. For example, whenever properties change our flag might change, and same for when the user identity changes. We can leverage this behaviour in our solution.
+
+### The Caching Problem
+
+It's not reasonable for us to cache the _results_ of flag evaluation for users because this blows up the size of the cache, doesn't work for new people we are seeing for the first time, and isn't flexible enough to leverage properties changing over time for users.
```

Comment:
```suggestion
It's not reasonable for us to cache the _results_ of flag evaluation for users because this:
1. blows up the size of the cache
2. doesn't work for new people we are seeing for the first time 
3. isn't flexible enough to leverage properties changing over time for users
```